supervisor/supervisor.go

```diff
@@ -0,0 +1,164 @@
+
+	signaled   bool
+	signaledMu sync.RWMutex
+
+	hupped   bool
+	huppedMu sync.RWMutex
+}
+
+func (sv *Supervisor) setSignaled(signaled bool) {
+	sv.signaledMu.Lock()
+	defer sv.signaledMu.Unlock()
+	sv.signaled = signaled
+}
+
+func (sv *Supervisor) getSignaled() bool {
+	sv.signaledMu.RLock()
+	defer sv.signaledMu.RUnlock()
+	return sv.signaled
+}
+
+func (sv *Supervisor) setHupped(hupped bool) {
+	sv.huppedMu.Lock()
+	defer sv.huppedMu.Unlock()
+	sv.hupped = hupped
+}
+
+func (sv *Supervisor) getHupped() bool {
+	sv.huppedMu.RLock()
+	defer sv.huppedMu.RUnlock()
+	return sv.hupped
+}
+
+func (sv *Supervisor) getCmd() *exec.Cmd {
+	sv.mu.RLock()
+	defer sv.mu.RUnlock()
+	return sv.cmd
+}
+
+func (sv *Supervisor) getStartAt() time.Time {
+	sv.mu.RLock()
+	defer sv.mu.RUnlock()
+	return sv.startAt
+}
+
+// If the child process dies within 30 seconds, it is regarded as launching failure
+// and terminate the process without crash recovery
+var spawnInterval = 30 * time.Second
+
+func (sv *Supervisor) launched() bool {
+	return sv.getCmd().Process != nil && time.Now().After(sv.getStartAt().Add(spawnInterval))
+}
+
+func (sv *Supervisor) buildCmd() *exec.Cmd {
+	argv := append(sv.Argv, "-child")
+	cmd := exec.Command(sv.Prog, argv...)
+	cmd.Stderr = os.Stderr
+	cmd.Stdout = os.Stdout
+	return cmd
+}
+
+func (sv *Supervisor) start() error {
+	sv.mu.Lock()
+	sv.setHupped(false)
+	defer sv.mu.Unlock()
+	sv.cmd = sv.buildCmd()
+	sv.startAt = time.Now()
+	return sv.cmd.Start()
+}
+
+func (sv *Supervisor) stop(sig os.Signal) error {
+	sv.setSignaled(true)
+	return sv.getCmd().Process.Signal(sig)
+}
+
+func (sv *Supervisor) configtest() error {
+	argv := append([]string{"configtest"}, sv.Argv...)
+	cmd := exec.Command(sv.Prog, argv...)
+	buf := &bytes.Buffer{}
+	cmd.Stderr = buf
+	err := cmd.Run()
+	if err != nil {
+		return fmt.Errorf("configtest failed: %s", buf.String())
+	}
+	return nil
+}
+
+func (sv *Supervisor) reload() error {
+	err := sv.configtest()
+	if err != nil {
+		return err
+	}
+	sv.setHupped(true)
+	return sv.getCmd().Process.Signal(syscall.SIGTERM)
+}
+
+func (sv *Supervisor) wait() (err error) {
+	for {
+		err = sv.cmd.Wait()
+		if sv.getSignaled() || (!sv.getHupped() && !sv.launched()) {
+			break
+		}
+		if err != nil {
+			logger.Warningf("mackerel-agent abnormally finished with following error and try to restart it: %s", err.Error())
+		}
+		err = sv.start()
+		if err != nil {
+			break
+		}
+	}
+	return
+}
+
+func (sv *Supervisor) handleSignal(ch <-chan os.Signal) {
+	for sig := range ch {
+		if sig == syscall.SIGHUP {
+			logger.Infof("receiving HUP, spawning a new mackerel-agent")
+			err := sv.reload()
+			if err != nil {
+				logger.Warningf("failed to reload: %s", err.Error())
+			}
+		} else {
+			sv.stop(sig)
+		}
+	}
+}
+
+// Supervise the mackerel-agent
+func (sv *Supervisor) Supervise(c chan os.Signal) error {
+	err := sv.start()
+	if err != nil {
+		return err
+	}
+	if c == nil {
+		c = make(chan os.Signal, 1)
```

Comment:
it's better to close `c` not to leak `go sv.handleSignal(c)` goroutine.

```go
defer close(c)
```

I also confirmed there is a leak with https://github.com/fortytw2/leaktest

Also, it's better to  document `c chan os.Signal` arg can be nil and if it's not nil, the caller has responsibility to close it. Or, we can just say "do not use this package as library", i guess?

e.g. 

```go
// Supervise starts a child mackerel-agent process and supervises it.
// 'c' can be nil and it's tipically nil. When you pass signal channel to this
// method, you have to close the channel to stop internal goroutine.
func (sv *Supervisor) Supervise(c chan os.Signal) error {
```